validator/client/src/main/java/tech/pegasys/teku/validator/client/doppelganger/DoppelgangerDetector.java

```diff
@@ -0,0 +1,297 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.TimeoutException;
+import java.util.stream.Collectors;
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import tech.pegasys.teku.api.migrated.ValidatorLivenessAtEpoch;
+import tech.pegasys.teku.bls.BLSPublicKey;
+import tech.pegasys.teku.infrastructure.async.AsyncRunner;
+import tech.pegasys.teku.infrastructure.async.Cancellable;
+import tech.pegasys.teku.infrastructure.async.SafeFuture;
+import tech.pegasys.teku.infrastructure.exceptions.ExceptionUtil;
+import tech.pegasys.teku.infrastructure.logging.StatusLogger;
+import tech.pegasys.teku.infrastructure.time.TimeProvider;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.spec.Spec;
+import tech.pegasys.teku.validator.api.ValidatorApiChannel;
+import tech.pegasys.teku.validator.beaconnode.GenesisDataProvider;
+
+public class DoppelgangerDetector {
+
+  private static final Logger LOG = LogManager.getLogger();
+  private final StatusLogger statusLog;
+  private final Duration checkDelay;
+  private final int maxEpochs;
+  private final Duration timeout;
+  private final AsyncRunner asyncRunner;
+  private final ValidatorApiChannel validatorApiChannel;
+  private final Spec spec;
+  private final TimeProvider timeProvider;
+  private final GenesisDataProvider genesisDataProvider;
+
+  public DoppelgangerDetector(
+      final AsyncRunner asyncRunner,
+      final ValidatorApiChannel validatorApiChannel,
+      final Spec spec,
+      final TimeProvider timeProvider,
+      final GenesisDataProvider genesisDataProvider,
+      final Duration checkDelay,
+      final Duration timeout,
+      final int maxEpochs) {
+    this(
+        STATUS_LOG,
+        asyncRunner,
+        validatorApiChannel,
+        spec,
+        timeProvider,
+        genesisDataProvider,
+        checkDelay,
+        timeout,
+        maxEpochs);
+  }
+
+  public DoppelgangerDetector(
+      final StatusLogger statusLog,
+      final AsyncRunner asyncRunner,
+      final ValidatorApiChannel validatorApiChannel,
+      final Spec spec,
+      final TimeProvider timeProvider,
+      final GenesisDataProvider genesisDataProvider,
+      final Duration checkDelay,
+      final Duration timeout,
+      final int maxEpochs) {
+    this.statusLog = statusLog;
+    this.asyncRunner = asyncRunner;
+    this.validatorApiChannel = validatorApiChannel;
+    this.spec = spec;
+    this.timeProvider = timeProvider;
+    this.genesisDataProvider = genesisDataProvider;
+    this.checkDelay = checkDelay;
+    this.timeout = timeout;
+    this.maxEpochs = maxEpochs;
+  }
+
+  public SafeFuture<Map<UInt64, BLSPublicKey>> performDoppelgangerDetection(
+      final Set<BLSPublicKey> pubKeys) {
+    if (pubKeys.isEmpty()) {
+      LOG.info("Skipping doppelganger detection: No public keys to check");
```

Comment:
doesn't look like it should be info?

Reply:
if it's once on startup I could live with it though...

Reply:
Yep it's not supposed to happen too often so I guess we can leave it like it is. Added a small detail though